### ISSUE TITLE
Misleading error message when trying to mutate outer variable captured with move

### ISSUE THREAD

**Issue by Diggsey** [rust-lang/rust#31871]
Code:

``` rust
type NonCopy = Box<u32>;

fn test<F: FnMut(NonCopy)>(f: &mut F) {
    f(NonCopy::new(21));
}

fn main() {
    let x = NonCopy::new(42);

    let mut closure = move |v| {
        println!("{}", x);
        x = v;
    };

    test(&mut closure);
    test(&mut closure);
}
```

Error:

```
<anon>:13:9: 13:14 error: cannot assign to captured outer variable in an `FnMut` closure
<anon>:13         x = v;
                  ^~~~~
```

The error message implies that mutating the outer variable is disallowed because this is a `FnMut` closure, or because mutating a capture is forbidden. However, the real problem is simply that `x` is not marked as mutable, and captured variables inherit their `mut`-status from the outer scope.

**Comment by GuillaumeGomez** [comment]
Interesting. I'll take a look when I have time (if no one did before me).

**Comment by birkenfeld** [comment]
Looking at the code, the error has to be parsed as
`cannot assign to (captured outer variable in an FnMut closure)`
and not as
`(cannot assign to captured outer variable) in an FnMut closure`
i.e. it the error is just "cannot assign to X", and has no intention of talking specifically about FnMuts.

**Comment by cyplo** [comment]
Seems that it's still an issue on rustc 1.15.0-nightly (d5814b03e 2016-11-23)

**Comment by estebank** [comment]
[Current output](https://play.rust-lang.org/?gist=08c3a699b746627e3a83c56c236d7172&version=nightly) points at the captured variable `x` with a suggestion to make it mutable:

```
error[E0594]: cannot assign to captured outer variable in an `FnMut` closure
  --> src/main.rs:12:9
   |
8  |     let x = NonCopy::new(42);
   |         - help: consider making `x` mutable: `mut x`
...
12 |         x = v;
   |         ^^^^^
```

**Comment by estebank** [comment]
Current output doesn't mention `FnMut` and applying the suggestion fixes the problem

```
error[E0594]: cannot assign to `x`, as it is not declared as mutable
  --> src/main.rs:12:9
   |
8  |     let x = NonCopy::new(42);
   |         - help: consider changing this to be mutable: `mut x`
...
12 |         x = v;
   |         ^ cannot assign
```
